signac/core/buffers.py

```diff
@@ -0,0 +1,132 @@
+
+logger = logging.getLogger(__name__)
+
+
+def _hash(blob):
+    """Calculate and return the md5 hash value for the file data."""
+    if blob is not None:
+        m = hashlib.md5()
+        m.update(blob)
+        return m.hexdigest()
+
+
+class FileBuffer:
+
+    _cache = get_cache()
+
+    @staticmethod
+    def _get_filemetadata(filename):
+        """Return metadata of JSON-file"""
+        try:
+            metadata = os.stat(filename)
+            return metadata.st_size, metadata.st_mtime
+        except OSError as error:
+            if error.errno != errno.ENOENT:
+                raise
+
+    def _store_to_buffer(self, filename, blob, store_hash=False):
+        if store_hash:
+            self._cache['HASHE::' + filename] = _hash(blob)
+        if filename not in self._cache:
+            if not (_in_buffered_mode() and _get_buffer_force_mode()):
+                # storing serialized metadata for redis-server
+                self._cache['METADATA::' + filename] = json.dumps(self._get_filemetadata(filename))
+            # adding filename to list
+            if isinstance(self._cache, dict):
+                if 'filenames' not in self._cache:
+                    self._cache['filenames'] = []
+                self._cache['filenames'].append(filename)
+            else:
+                self._cache.lpush('filenames', filename)
+        self._cache[filename] = blob
+
+    @classmethod
+    def _pop_from_cache(cls, key, decode=False):
+        # Redis client does not have `pop`.
+        val = cls._cache[key]
+        del cls._cache[key]
+        return val.decode() if decode and isinstance(val, bytes) else val
+
+    @staticmethod
+    @abstractmethod
+    def _write_to_file(filename, blob):
+        """Write data to file."""
+
+    @classmethod
+    def _flush(cls, filename):
+        """Write the data from buffer to the file. Return error if any."""
+        blob = cls._pop_from_cache(filename)
+
+        if not _get_buffer_force_mode():
+            meta = json.loads(cls._pop_from_cache('METADATA::' + filename))
+
+        # if hash stored and hash of data is same then we skip the sync
+        hash_key = 'HASHE::' + filename
+        if not (hash_key in cls._cache and _hash(blob) == cls._pop_from_cache(hash_key,
+                                                                              decode=True)):
+            # compare the metadata
+            if (not _get_buffer_force_mode()) and meta is not None:
+                if cls._get_filemetadata(filename) != tuple(meta):
+                    return 'File appears to have been externally modified.'
+            # Sync the data to underlying backend
+            try:
+                cls._write_to_file(filename, blob)
+            except OSError as error:
+                # if sync fails return error
+                logger.error(str(error))
+                return error
+
+    @classmethod
+    def _flush_buffer(cls):
+        """Flush the data in JSON-buffer.
```

Comment:
At the end of this method, is it expected that the key `filenames` is **set as an empty list** (`assert cls._cache['filenames'] == []`), or that it is **unset** (`assert 'filenames' not in cls._cache`)? It is not clear to me which is desired. Currently I would anticipate (without testing it) that the `dict` cache key is **unset** while the Redis cache key is **an empty list**. Please clarify the intended behavior in the docstring and verify both caches.

Reply:
Redis client removes the key if all the data is removed from the list. So, I expected `filenames` to unset.